File: app/src/main/java/com/nirbhor/app/ui/screens/add/AddTimingScreen.kt
package com.nirbhor.app.ui.screens

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.FlowRow
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.navigationBarsPadding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.foundation.clickable
import androidx.compose.foundation.border
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Brightness5
import androidx.compose.material.icons.filled.DarkMode
import androidx.compose.material.icons.filled.WbSunny
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.unit.dp
import com.nirbhor.app.domain.FoodRelation
import com.nirbhor.app.domain.Frequency
import com.nirbhor.app.domain.LocalAddDraft
import com.nirbhor.app.domain.TimeBlock
import com.nirbhor.app.navigation.NavActions
import com.nirbhor.app.ui.components.PrimaryButton
import com.nirbhor.app.ui.components.QuickChip
import com.nirbhor.app.ui.components.SegmentedControl
import com.nirbhor.app.ui.components.SelectableRow
import com.nirbhor.app.ui.components.TintPanel
import com.nirbhor.app.ui.i18n.tr
import com.nirbhor.app.ui.theme.Dimens
import com.nirbhor.app.ui.theme.NirbhorTheme

/** When to take (3c). No clock picker — meal tokens resolve to default times. Multi-select. */
@Composable
fun AddTimingScreen(actions: NavActions) {
    val colors = NirbhorTheme.colors
    val draft = LocalAddDraft.current

    fun toggle(token: String) {
        draft.timeTokens = if (token in draft.timeTokens) draft.timeTokens - token else draft.timeTokens + token
        // Keep resolved times aligned to selected tokens in block order.
        val ordered = TimeBlock.entries.filter { it.token in draft.timeTokens }
        draft.timeTokens = ordered.map { it.token }
        draft.resolvedTimes = ordered.map { "%02d:00".format(it.defaultHour) }
    }

    Column(Modifier.fillMaxSize().background(colors.paper)) {
        AddFlowHeader(step = 1, onBack = actions::back)
        Column(
            Modifier.fillMaxSize().verticalScroll(rememberScrollState()).navigationBarsPadding().padding(horizontal = Dimens.screenPadding, vertical = 8.dp),
            verticalArrangement = Arrangement.spacedBy(Dimens.groupGap),
        ) {
            Text(tr("কখন খাবেন?", "When do you take it?"), style = NirbhorTheme.type.titleHero, color = colors.ink)
            Column(verticalArrangement = Arrangement.spacedBy(Dimens.cardGap)) {
                TimeChoice(Icons.Filled.WbSunny, tr("সকাল", "Morning"), TimeBlock.MORNING.token in draft.timeTokens) { toggle(TimeBlock.MORNING.token) }
                TimeChoice(Icons.Filled.Brightness5, tr("দুপুর", "Afternoon"), TimeBlock.NOON.token in draft.timeTokens) { toggle(TimeBlock.NOON.token) }
                TimeChoice(Icons.Filled.DarkMode, tr("রাত", "Night"), TimeBlock.NIGHT.token in draft.timeTokens) { toggle(TimeBlock.NIGHT.token) }
            }

            Text(tr("কত ঘন ঘন?", "How often?"), style = NirbhorTheme.type.header, color = colors.ink)
            SegmentedControl(
                options = listOf(tr("প্রতিদিন", "Daily"), tr("একদিন পরপর", "Alternate"), tr("দিন বেছে", "Select days")),
                selectedIndex = when (draft.frequency) {
                    Frequency.DAILY -> 0
                    Frequency.ALTERNATE -> 1
                    Frequency.WEEKDAYS, Frequency.WEEKLY -> 2
                },
                onSelect = { index ->
                    draft.frequency = when (index) {
                        1 -> Frequency.ALTERNATE
                        2 -> Frequency.WEEKLY
                        else -> Frequency.DAILY
                    }
                    if (index == 2 && draft.weekdaysMask == 0) {
                        draft.weekdaysMask = com.nirbhor.app.domain.DoseScheduler.weekdayBit(java.time.LocalDate.now().dayOfWeek)
                    }
                },
                modifier = Modifier.fillMaxWidth(),
            )
            if (draft.frequency == Frequency.WEEKLY || draft.frequency == Frequency.WEEKDAYS) {
                val dayLabels = listOf(
                    tr("সোম", "Mon"), tr("মঙ্গল", "Tue"), tr("বুধ", "Wed"), tr("বৃহঃ", "Thu"),
                    tr("শুক্র", "Fri"), tr("শনি", "Sat"), tr("রবি", "Sun"),
                )
                FlowRow(
                    modifier = Modifier.fillMaxWidth(),
                    horizontalArrangement = Arrangement.spacedBy(8.dp),
                    verticalArrangement = Arrangement.spacedBy(8.dp),
                    maxItemsInEachRow = 4,
                ) {
                    dayLabels.forEachIndexed { index, label ->
                        val bit = 1 shl index
                        QuickChip(
                            label = label,
                            selected = draft.weekdaysMask and bit != 0,
                            onClick = {
                                val next = draft.weekdaysMask xor bit
                                if (next != 0) draft.weekdaysMask = next
                            },
                            modifier = Modifier.width(76.dp),
                            height = 48.dp,
                        )
                    }
                }
            }

            Text(tr("খাবারের আগে না পরে?", "Before or after food?"), style = NirbhorTheme.type.header, color = colors.ink)
            Row(Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.spacedBy(Dimens.cardGap)) {
                FoodCard(tr("আগে / খালি পেটে", "Before / empty"), draft.foodRelation == FoodRelation.BEFORE, Modifier.weight(1f)) { draft.foodRelation = FoodRelation.BEFORE }
                FoodCard(tr("পরে / ভরা পেটে", "After / full"), draft.foodRelation == FoodRelation.AFTER, Modifier.weight(1f)) { draft.foodRelation = FoodRelation.AFTER }
            }

            if (draft.timeTokens.isNotEmpty()) {
                val bangla = com.nirbhor.app.ui.i18n.LocalIsBangla.current
                val is24 = com.nirbhor.app.ui.i18n.LocalIs24Hour.current
                val sep = tr(" আর ", " and ")
                val times = draft.resolvedTimes.joinToString(sep) { hhmm ->
                    val h = hhmm.substringBefore(":").toInt()
                    com.nirbhor.app.ui.i18n.Numerals.time(h, 0, bangla, is24)
                }
                TintPanel(background = colors.calmSoft) {
                    Text(
                        tr("তাহলে অ্যালার্ম বাজবে $times-এ। সময় পছন্দ না হলে পরে বদলে নিতে পারবেন।",
                            "Alarms will ring at $times. You can change the times later."),
                        style = NirbhorTheme.type.body, color = colors.calmD,
                    )
                }
            }
            PrimaryButton(tr("পরের ধাপ", "Next"), actions::addQuantity, height = 64.dp, enabled = draft.timeTokens.isNotEmpty(), modifier = Modifier.fillMaxWidth())
            Spacer(Modifier.height(8.dp))
        }
    }
}

@Composable
private fun TimeChoice(icon: ImageVector, label: String, selected: Boolean, onClick: () -> Unit) {
    val colors = NirbhorTheme.colors
    SelectableRow(
        title = label, selected = selected, onClick = onClick, height = 88.dp,
        leading = {
            Icon(icon, contentDescription = null, tint = if (selected) colors.paper else colors.ink2, modifier = Modifier.size(34.dp))
        },
    )
}

@Composable
private fun FoodCard(label: String, selected: Boolean, modifier: Modifier, onClick: () -> Unit) {
    val colors = NirbhorTheme.colors
    val shape = RoundedCornerShape(14.dp)
    Box(
        modifier.height(76.dp).clip(shape)
            .background(if (selected) colors.calm else colors.card)
            .border(if (selected) 2.dp else 1.5.dp, if (selected) colors.calm else colors.line, shape)
            .clickable(onClick = onClick).padding(14.dp),
        contentAlignment = Alignment.CenterStart,
    ) {
        Text(label, style = NirbhorTheme.type.cardTitleSecondary, color = if (selected) colors.paper else colors.ink)
    }
}

File: lib/ui/screens/add_timing_screen.dart
import 'package:flutter/material.dart';

import '../../domain/dose_scheduler.dart';
import '../../domain/models.dart';
import '../../i18n/numerals.dart';
import '../../navigation/nav_actions.dart';
import '../../theme/theme.dart';
import '../components/buttons.dart';
import '../components/controls.dart';
import '../components/surfaces.dart';
import 'add_flow_common.dart';

/// When to take it (3c) — the time blocks, the frequency, and the meal relation.
class AddTimingScreen extends StatelessWidget {
  const AddTimingScreen({super.key});

  void _toggle(BuildContext context, String token) {
    final draft = context.draft;
    draft.update(() {
      final tokens = List.of(draft.timeTokens);
      if (!tokens.remove(token)) tokens.add(token);
      // Keep resolved times aligned to selected tokens in block order.
      final ordered = TimeBlock.values.where((b) => tokens.contains(b.token)).toList();
      draft.timeTokens = ordered.map((b) => b.token).toList();
      draft.resolvedTimes = ordered.map(DoseScheduler.blockDefaultTime).toList();
    });
  }

  @override
  Widget build(BuildContext context) {
    final colors = context.colors;
    final draft = context.draft;

    return Scaffold(
      backgroundColor: colors.paper,
      body: ListenableBuilder(
        listenable: draft,
        builder: (context, _) {
          final selectedIndex = switch (draft.frequency) {
            Frequency.daily => 0,
            Frequency.alternate => 1,
            Frequency.weekdays || Frequency.weekly => 2,
          };
          final showDays =
              draft.frequency == Frequency.weekly || draft.frequency == Frequency.weekdays;

          final sep = context.tr(' আর ', ' and ');
          final times = draft.resolvedTimes
              .map((hhmm) {
                final parsed = DoseScheduler.parseHhmm(hhmm) ?? (8, 0);
                return Numerals.time(parsed.$1, parsed.$2, context.isBangla, context.is24Hour);
              })
              .join(sep);

          return Column(
            children: [
              AddFlowHeader(step: 1, onBack: context.nav.back),
              Expanded(
                child: SingleChildScrollView(
                  padding: const EdgeInsets.symmetric(
                    horizontal: Dimens.screenPadding,
                    vertical: 8,
                  ),
                  child: SafeArea(
                    top: false,
                    child: Column(
                      crossAxisAlignment: CrossAxisAlignment.stretch,
                      children: [
                        Text(
                          context.tr('কখন খাবেন?', 'When do you take it?'),
                          style: context.type.titleHero.copyWith(color: colors.ink),
                        ),
                        const SizedBox(height: Dimens.groupGap),
                        _TimeChoice(
                          icon: Icons.wb_sunny,
                          label: context.tr('সকাল', 'Morning'),
                          selected: draft.timeTokens.contains(TimeBlock.morning.token),
                          onTap: () => _toggle(context, TimeBlock.morning.token),
                        ),
                        const SizedBox(height: Dimens.cardGap),
                        _TimeChoice(
                          icon: Icons.brightness_5,
                          label: context.tr('দুপুর', 'Afternoon'),
                          selected: draft.timeTokens.contains(TimeBlock.noon.token),
                          onTap: () => _toggle(context, TimeBlock.noon.token),
                        ),
                        const SizedBox(height: Dimens.cardGap),
                        _TimeChoice(
                          icon: Icons.dark_mode,
                          label: context.tr('রাত', 'Night'),
                          selected: draft.timeTokens.contains(TimeBlock.night.token),
                          onTap: () => _toggle(context, TimeBlock.night.token),
                        ),
                        const SizedBox(height: Dimens.groupGap),

                        Text(
                          context.tr('কত ঘন ঘন?', 'How often?'),
                          style: context.type.header.copyWith(color: colors.ink),
                        ),
                        const SizedBox(height: Dimens.groupGap),
                        SegmentedControl(
                          options: [
                            context.tr('প্রতিদিন', 'Daily'),
                            context.tr('একদিন পরপর', 'Alternate'),
                            context.tr('দিন বেছে', 'Select days'),
                          ],
                          selectedIndex: selectedIndex,
                          onSelect: (index) => draft.update(() {
                            draft.frequency = switch (index) {
                              1 => Frequency.alternate,
                              2 => Frequency.weekly,
                              _ => Frequency.daily,
                            };
                            // Landing on "select days" with nothing ticked would silently mean
                            // "never" — start from today.
                            if (index == 2 && draft.weekdaysMask == 0) {
                              draft.weekdaysMask =
                                  DoseScheduler.weekdayBit(DateTime.now().weekday);
                            }
                          }),
                        ),
                        if (showDays) ...[
                          const SizedBox(height: Dimens.groupGap),
                          Wrap(
                            spacing: 8,
                            runSpacing: 8,
                            children: [
                              for (var index = 0; index < 7; index++)
                                SizedBox(
                                  width: 76,
                                  child: QuickChip(
                                    label: [
                                      context.tr('সোম', 'Mon'),
                                      context.tr('মঙ্গল', 'Tue'),
                                      context.tr('বুধ', 'Wed'),
                                      context.tr('বৃহঃ', 'Thu'),
                                      context.tr('শুক্র', 'Fri'),
                                      context.tr('শনি', 'Sat'),
                                      context.tr('রবি', 'Sun'),
                                    ][index],
                                    height: 48,
                                    multiSelect: true,
                                    selected: draft.weekdaysMask & (1 << index) != 0,
                                    onTap: () => draft.update(() {
                                      final next = draft.weekdaysMask ^ (1 << index);
                                      // Never let the last day be unticked into "never".
                                      if (next != 0) draft.weekdaysMask = next;
                                    }),
                                  ),
                                ),
                            ],
                          ),
                        ],
                        const SizedBox(height: Dimens.groupGap),

                        Text(
                          context.tr('খাবারের আগে না পরে?', 'Before or after food?'),
                          style: context.type.header.copyWith(color: colors.ink),
                        ),
                        const SizedBox(height: Dimens.groupGap),
                        Row(
                          children: [
                            Expanded(
                              child: _FoodCard(
                                label: context.tr('আগে / খালি পেটে', 'Before / empty'),
                                selected: draft.foodRelation == FoodRelation.before,
                                onTap: () => draft
                                    .update(() => draft.foodRelation = FoodRelation.before),
                              ),
                            ),
                            const SizedBox(width: Dimens.cardGap),
                            Expanded(
                              child: _FoodCard(
                                label: context.tr('পরে / ভরা পেটে', 'After / full'),
                                selected: draft.foodRelation == FoodRelation.after,
                                onTap: () => draft
                                    .update(() => draft.foodRelation = FoodRelation.after),
                              ),
                            ),
                          ],
                        ),

                        if (draft.timeTokens.isNotEmpty) ...[
                          const SizedBox(height: Dimens.groupGap),
                          TintPanel(
                            background: colors.calmSoft,
                            child: Text(
                              context.tr(
                                'তাহলে অ্যালার্ম বাজবে $times-এ। সময় পছন্দ না হলে পরে বদলে নিতে পারবেন।',
                                'Alarms will ring at $times. You can change the times later.', hi: 'अलार्म $times पर बजेंगे। समय बाद में बदल सकते हैं।', es: 'Las alarmas sonarán a las $times. Puede cambiar las horas después.',
                              ),
                              style: context.type.body.copyWith(color: colors.calmD),
                            ),
                          ),
                        ],
                        const SizedBox(height: Dimens.groupGap),
                        PrimaryButton(
                          text: context.tr('পরের ধাপ', 'Next'),
                          onPressed: context.nav.addQuantity,
                          enabled: draft.timeTokens.isNotEmpty,
                        ),
                        const SizedBox(height: 8),
                      ],
                    ),
                  ),
                ),
              ),
            ],
          );
        },
      ),
    );
  }
}

class _TimeChoice extends StatelessWidget {
  const _TimeChoice({
    required this.icon,
    required this.label,
    required this.selected,
    required this.onTap,
  });

  final IconData icon;
  final String label;
  final bool selected;
  final VoidCallback onTap;

  @override
  Widget build(BuildContext context) {
    final colors = context.colors;
    return SelectableRow(
      title: label,
      selected: selected,
      onTap: onTap,
      leading: Icon(icon, size: 34, color: selected ? colors.paper : colors.ink2),
    );
  }
}

class _FoodCard extends StatelessWidget {
  const _FoodCard({required this.label, required this.selected, required this.onTap});

  final String label;
  final bool selected;
  final VoidCallback onTap;

  @override
  Widget build(BuildContext context) {
    final colors = context.colors;
    return Semantics(
      selected: selected,
      inMutuallyExclusiveGroup: true,
      child: GestureDetector(
        behavior: HitTestBehavior.opaque,
        onTap: onTap,
        child: Container(
          height: 76,
          alignment: Alignment.centerLeft,
          padding: const EdgeInsets.all(14),
          decoration: BoxDecoration(
            color: selected ? colors.calm : colors.card,
            borderRadius: BorderRadius.circular(14),
            border: Border.all(
              color: selected ? colors.calm : colors.line,
              width: selected ? 2 : 1.5,
            ),
          ),
          child: Text(
            label,
            style: context.type.cardTitleSecondary
                .copyWith(color: selected ? colors.paper : colors.ink),
          ),
        ),
      ),
    );
  }
}
